SimGeneral/PreMixingModule/plugins/PreMixingTrackingParticleWorker.cc

```diff
@@ -0,0 +1,156 @@
+    void addSignals(edm::Event const& iEvent, edm::EventSetup const& iSetup) override;
+    void addPileups(int bcr, edm::EventPrincipal const& ep, int eventNr, edm::EventSetup const& iSetup, edm::ModuleCallingContext const *mcc) override;
+    void put(edm::Event& iEvent, edm::EventSetup const& iSetup, std::vector<PileupSummaryInfo> const& ps, int bunchSpacing) override;
+
+  private:
+    void add(const std::vector<TrackingParticle>& particles, const std::vector<TrackingVertex>& vertices);
+    
+    edm::EDGetTokenT<std::vector<TrackingParticle> > TrackSigToken_;  // Token to retrieve information  
+    edm::EDGetTokenT<std::vector<TrackingVertex> > VtxSigToken_;  // Token to retrieve information  
+
+    edm::InputTag TrackingParticlePileInputTag_ ;    // InputTag for pileup tracks
+
+    std::string TrackingParticleCollectionDM_; // secondary name to be given to new TrackingParticle
+
+
+    std::unique_ptr<std::vector<TrackingParticle>> NewTrackList_;
+    std::unique_ptr<std::vector<TrackingVertex>> NewVertexList_;
+    TrackingParticleRefProd TrackListRef_;
+    TrackingVertexRefProd VertexListRef_;
+  };
+
+  PreMixingTrackingParticleWorker::PreMixingTrackingParticleWorker(const edm::ParameterSet& ps, edm::ProducerBase& producer, edm::ConsumesCollector && iC):
+    TrackSigToken_(iC.consumes<std::vector<TrackingParticle> >(ps.getParameter<edm::InputTag>("labelSig"))),
+    VtxSigToken_  (iC.consumes<std::vector<TrackingVertex>   >(ps.getParameter<edm::InputTag>("labelSig"))),
+    TrackingParticlePileInputTag_(ps.getParameter<edm::InputTag>("pileInputTag")),
+    TrackingParticleCollectionDM_(ps.getParameter<std::string>("collectionDM"))
+  {
+    producer.produces< std::vector<TrackingParticle> >(TrackingParticleCollectionDM_);
+    producer.produces< std::vector<TrackingVertex> >(TrackingParticleCollectionDM_);
+  }
+
+
+  void PreMixingTrackingParticleWorker::initializeEvent(edm::Event const& iEvent, edm::EventSetup const& iSetup) {
+    NewTrackList_ = std::make_unique<std::vector<TrackingParticle>>();
+    NewVertexList_ = std::make_unique<std::vector<TrackingVertex>>();
+
+    // need RefProds in order to re-key the particle<->vertex refs
+    // TODO: try to remove const_cast, requires making Event non-const in BMixingModule::initializeEvent
+    TrackListRef_  = const_cast<edm::Event&>(iEvent).getRefBeforePut< std::vector<TrackingParticle> >(TrackingParticleCollectionDM_);
+    VertexListRef_ = const_cast<edm::Event&>(iEvent).getRefBeforePut< std::vector<TrackingVertex> >(TrackingParticleCollectionDM_);    
+  }
+
+  void PreMixingTrackingParticleWorker::addSignals(edm::Event const& iEvent, edm::EventSetup const& iSetup) {
+    edm::Handle<std::vector<TrackingParticle>> tracks;
+    iEvent.getByToken(TrackSigToken_, tracks);
+
+    edm::Handle<std::vector<TrackingVertex>> vtxs;
+    iEvent.getByToken(VtxSigToken_, vtxs);
+
+    if(tracks.isValid() && vtxs.isValid()) {
```

Comment:
Personally I'm not happy with these checks in cases where the `Event` is expected to always contain the products as I suppose is the case here (because it delays catching mistakes). Here (and for `SimDigiLink`s) I chose to keep them, because otherwise I would have had to patch `ConfigBuilder` similarly to what it does for `MixingModule` when `DIGI:pdigi_valid` is specified
https://github.com/cms-sw/cmssw/blob/997d1ea4289d52d453446b2d378a44e2010f08a2/Configuration/Applications/python/ConfigBuilder.py#L1414-L1415
and I didn't want to think how do that cleanly for `prepare_DATAMIX`.

If there is no strong reason to keep the checks, is the only option to patch `ConfigBuilder`, or would it make sense e.g. to use a `cms.Modifier` to do the customizations for validation? (that is,
* change `pdigi` -> `pdigi_valid`
* customize `process.mix.digitizers` to the validation variant
* customize `process.mixData.workers` to the validation variant

)

Reply:
I'm not sure what the best way is to deal with ConfigBuilder.  That code is already such a mess.

Reply:
indeed, what @makortel suggests could reduce rather than increase complexity..

Reply:
I created an issue #22494 to remind to test the Modifier approach.